### PR TITLE
sql/schemachanger: Enable adding/dropping path of check constraint

### PR DESCRIPTION
**Pull request by Xiang-Gu**
Previously, we emit unimplemented operations on the adding and dropping
path of check constraint and constraint name. This PR enables this feature
as it will unblock certain DDL statement in the declarative schema changer
such as `ALTER PRIMARY KEY USING HASH`.

The last two commits are the "meaty" commits:
1). Implemented the logic to validate a check constraint in `backfill.go` 
     and plumb it in to the Validator used in declarative schema changer
     for unique index and check constraint validation.

2). Introduced a new status -- `VALIDATING`, and now the 
     adding/dropping path of a check constraint is
    ABSENT <--> VALIDATING <--> VALIDATED <--> PUBLIC

Release note: None

### CODE REVIEW

**Comment by Xiang-Gu** [comment]
RFAL!

File: pkg/sql/schemachanger/scpb/scpb.proto [hunk 1]
```diff
@@ -53,6 +53,10 @@ enum Status {
   // Intermediate states on the transient index dropping path
   TRANSIENT_DELETE_ONLY = 13;
   TRANSIENT_WRITE_ONLY = 14;
+
+  // Intermediate states on check constraint dropping and adding path.
+  Check_Constraint_Validating = 16;
+  Check_Constraint_Validated = 17;
```
Comment (postamar) [inline pkg/sql/schemachanger/scpb/scpb.proto:59]: We'll need similar states for other constraints so I'm not sure I like these very much. Can we reuse existing states? What about reusing VALIDATED, and using WRITE_ONLY for "validating"? From what I understand, in that state, we apply the constraint only on stuff we write to the table. Please correct me if I'm wrong.

File: pkg/sql/schemachanger/scexec/scmutationexec/references.go [hunk 2]
```diff
@@ -89,6 +90,80 @@ func (m *visitor) RemoveCheckConstraint(ctx context.Context, op scop.RemoveCheck
...
+
+	return nil
+}
+
+func (m *visitor) MakeDroppedCheckConstraintValidated(
+	ctx context.Context, constraint scop.MakeDroppedCheckConstraintValidated,
+) error {
+	// TODO (xiang): implement this
+	return nil
+}
```
Comment (postamar) [inline pkg/sql/schemachanger/scexec/scmutationexec/references.go:165]: Does this op even need to exist? If it's not going anything, I'd rather it didn't.

File: pkg/sql/schemachanger/scplan/internal/rules/helpers.go [hunk 3]
```diff
@@ -255,6 +255,14 @@ func isColumn(e scpb.Element) bool {
 	return false
 }
 
+func isCheckConstraint(e scpb.Element) bool {
+	switch e.(type) {
+	case *scpb.CheckConstraint:
+		return true
+	}
+	return false
+}
```
Comment (postamar) [inline pkg/sql/schemachanger/scplan/internal/rules/helpers.go:264]: This function is only ever going to map to one element type so I don't think it's very useful.

**Review by postamar — COMMENTED** [review]
This is off to a good start. I can't find much wrong with it, but you still have a lot of failing tests. Fixing those are going to be hard, it always is.

**Review by Xiang-Gu — COMMENTED** [review]
I've address the comments. I've discovered and fixed a few failed tests. Would appreciate a review again!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85836#-:-N9J3I7A1O36OqopmLTo:bp4s1v1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @postamar, and @rhu713)

___
*[`pkg/sql/schemachanger/scexec/scmutationexec/references.go` line 165 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85836#-N92tLDkxBwcNJBTh2tu-r5-165:-N9IeleTCbQJ2yugWquV:ba9oxc1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd7188cb3621c2199c95397baa4345d91b5ea/pkg/sql/schemachanger/scexec/scmutationexec/references.go#L165)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Does this op even need to exist? If it's not going anything, I'd rather it didn't.
</blockquote></details>

sorry, I forgot to implement it. I've implemented it now.
___
*[`pkg/sql/schemachanger/scpb/scpb.proto` line 59 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85836#-N92tLDkxBwcNJBTh2tx-r5-59:-N9IgZcK8irWgyvwaMjf:bkyqd2v) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd7188cb3621c2199c95397baa4345d91b5ea/pkg/sql/schemachanger/scpb/scpb.proto#L59)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

We'll need similar states for other constraints so I'm not sure I like these very much. Can we reuse existing states? What about reusing VALIDATED, and using WRITE_ONLY for "validating"? From what I understand, in that state, we apply the constraint only on stuff we write to the table. Please correct me if I'm wrong.
</blockquote></details>

I think you are right! @ajwerner please correct us if we are wrong.
A check constraint in "validating" means it will be enforced for new writes, so I do think "write_only" is an appropriate status to substitute for.
___
*[`pkg/sql/schemachanger/scplan/internal/rules/helpers.go` line 264 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85836#-N92tLDkxBwcNJBTh2tz-r5-264:-N9Ih3XtEY-fRupUvoES:bam18xm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd7188cb3621c2199c95397baa4345d91b5ea/pkg/sql/schemachanger/scplan/internal/rules/helpers.go#L264)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This function is only ever going to map to one element type so I don't think it's very useful.
</blockquote></details>

This function resembles `isColumn` above and is used to neatly implement the `isSubjectTo2VersionInvariant` function above.

**Review by ajwerner — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/85836)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar and @rhu713)

___
*[`pkg/sql/schemachanger/scpb/scpb.proto` line 59 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85836#-N92tLDkxBwcNJBTh2tx-r5-59:-N9J5CMO1a8K7oM7NknQ:b-6hcyf3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd7188cb3621c2199c95397baa4345d91b5ea/pkg/sql/schemachanger/scpb/scpb.proto#L59)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

I think you are right! @ajwerner please correct us if we are wrong.
A check constraint in "validating" means it will be enforced for new writes, so I do think "write_only" is an appropriate status to substitute for.
</blockquote></details>

You should use the capitalization convention we have elsewhere. I also don't like the prefixing. Just re-use `VALIDATED`? I also agree that `WRITE_ONLY` is very similar in the we enforce things like `NOT NULL` and uniqueness in `WRITE_ONLY`. I wouldn't find `VALIDATING` to be offensive though if you wanted to add it. 
___
*[`pkg/sql/schemachanger/scplan/internal/rules/helpers.go` line 264 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/85836#-N92tLDkxBwcNJBTh2tz-r5-264:-N9J5dqf3lVME2AILw3x:b-7920qb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ad9bd7188cb3621c2199c95397baa4345d91b5ea/pkg/sql/schemachanger/scplan/internal/rules/helpers.go#L264)):*
<details><summary><i>Previously, Xiang-Gu (Xiang Gu) wrote…</i></summary><blockquote>

This function resembles `isColumn` above and is used to neatly implement the `isSubjectTo2VersionInvariant` function above.
</blockquote></details>

Can we change both of them to be simpler like:

```go
_, ok := e.(*CheckConstraint)
return ok
```

**Comment by postamar** [comment]
Let's table this until the next release. This was a valiant and impressive effort, but time has run out. 

If it's any consolation, we experienced something similar in the last cycle with ADD COLUMN. We thought we were so close! Retrospectively, I'm glad we gave that the proper treatment it deserved in the subsequent (current) release cycle. I'm having similar gut feelings here for constraints in general. 

Also, who knows what unknown problems the hash-sharding will reveal, and that's all still to be done.

**Comment by Xiang-Gu** [comment]
I'm going to close this in favor of a sequence of smaller PRs, as tracked in issue https://github.com/cockroachdb/cockroach/issues/89665

**Pull request closed without merging**